File: src/scripts/_lib/update_check.py
"""Daily update-check banner for the ``agent-config`` dispatcher.

Phase 2 of road-to-portable-runtime-and-update-check. Pure functions:
``check_for_update()`` decides whether a banner should be emitted and
returns the banner string (or ``None``). The dispatcher prints the
returned string to ``stderr`` after the subcommand finishes — never
delaying the work, never prompting.

Design constraints (see roadmap P2):

- Stdlib only (no new deps); the package's Python floor is stdlib-only.
- 1 s hard timeout on the registry call; network failure is silent.
- 24 h cadence gated by ``~/.event4u/agent-config/update-check.json``
  (legacy ``~/.config/agent-config/update-check.json`` is read once as
  a fallback so the cadence is not reset on the first run after the
  Phase-1 namespace migration).
- Suppress in CI, on non-TTY stdout, when ``AGENT_CONFIG_NO_UPDATE_CHECK=1``,
  or when ``update_check.enabled: false`` in settings.
- State file mode is ``0600``.

The dispatcher is the only call site. Tests mock ``now``, the state
path, and ``fetch_latest_from_npm`` to cover every branch.
"""
from __future__ import annotations

import json
import os
import sys
import tempfile
import urllib.error
import urllib.request
from datetime import datetime, timedelta, timezone
from pathlib import Path
from typing import Optional

from scripts._lib import user_global_paths

PACKAGE_NAME = "@event4u/agent-config"
NPM_REGISTRY_URL = f"https://registry.npmjs.org/{PACKAGE_NAME}/latest"
FETCH_TIMEOUT_S = 1.0
CHECK_WINDOW = timedelta(hours=24)

STATE_FILENAME = "update-check.json"

#: Canonical write target. Reads are routed via
#: :func:`_resolve_state_path` with a read-fallback to the legacy
#: ``~/.config/agent-config/update-check.json`` so a fresh install
#: under the new namespace does not lose the 24 h cadence window
#: established by a pre-2.4 install.
DEFAULT_STATE_PATH = user_global_paths.write_target(STATE_FILENAME)


def _resolve_state_path() -> Path:
    """Return the active state path, preferring the new namespace."""
    found = user_global_paths.resolve_with_fallback(STATE_FILENAME)
    if found is not None:
        return found
    return DEFAULT_STATE_PATH


def _now_utc() -> datetime:
    return datetime.now(timezone.utc)


def fetch_latest_from_npm(
    *,
    timeout: float = FETCH_TIMEOUT_S,
    url: str = NPM_REGISTRY_URL,
) -> Optional[str]:
    """Return the ``latest`` dist-tag version, or ``None`` on any failure.

    Hard 1 s timeout. Any exception (network, JSON, missing key) yields
    ``None`` — the update check is best-effort.
    """
    try:
        req = urllib.request.Request(
            url,
            headers={"Accept": "application/json", "User-Agent": "agent-config-update-check"},
        )
        with urllib.request.urlopen(req, timeout=timeout) as resp:  # noqa: S310 — fixed URL
            payload = json.load(resp)
        version = payload.get("version")
        if isinstance(version, str) and version.strip():
            return version.strip()
    except (urllib.error.URLError, TimeoutError, ValueError, OSError, json.JSONDecodeError):
        return None
    return None


def _read_state(path: Path) -> dict:
    try:
        with path.open("r", encoding="utf-8") as fh:
            data = json.load(fh)
        if isinstance(data, dict):
            return data
    except (OSError, ValueError, json.JSONDecodeError):
        pass
    return {}


def _write_state(path: Path, payload: dict) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    fd, tmp = tempfile.mkstemp(prefix=".update-check-", dir=str(path.parent))
    try:
        with os.fdopen(fd, "w", encoding="utf-8") as fh:
            json.dump(payload, fh, indent=2, sort_keys=True)
        os.chmod(tmp, 0o600)
        os.replace(tmp, path)
    except Exception:
        try:
            os.unlink(tmp)
        except OSError:
            pass
        raise


def _should_check(state: dict, now: datetime) -> bool:
    last = state.get("last_check_utc")
    if not isinstance(last, str):
        return True
    try:
        last_dt = datetime.fromisoformat(last.replace("Z", "+00:00"))
    except ValueError:
        return True
    if last_dt.tzinfo is None:
        last_dt = last_dt.replace(tzinfo=timezone.utc)
    return (now - last_dt) >= CHECK_WINDOW


def _format_banner(latest: str, installed: str) -> str:
    return (
        f"ℹ️  agent-config {latest} available (you have {installed}).\n"
        f"    Update: npx {PACKAGE_NAME} update"
    )


def _is_newer(latest: str, installed: str) -> bool:
    def _parse(v: str) -> tuple:
        parts = v.lstrip("v").split("-", 1)[0].split(".")
        out = []
        for p in parts[:3]:
            try:
                out.append(int(p))
            except ValueError:
                out.append(0)
        while len(out) < 3:
            out.append(0)
        return tuple(out)

    return _parse(latest) > _parse(installed)


def check_for_update(
    installed_version: str,
    *,
    now: Optional[datetime] = None,
    state_path: Optional[Path] = None,
    env: Optional[dict] = None,
    is_tty: Optional[bool] = None,
    settings_enabled: bool = True,
    fetcher=fetch_latest_from_npm,
) -> Optional[str]:
    """Decide whether to show an update banner. Pure (modulo state file).

    Returns the banner string or ``None``. ``None`` covers every
    suppression branch (CI, non-TTY, opt-out, within 24 h, network
    failure, no update available).
    """
    env = env if env is not None else os.environ
    if env.get("AGENT_CONFIG_NO_UPDATE_CHECK") == "1":
        return None
    if env.get("CI") in {"1", "true"} or env.get("GITHUB_ACTIONS") == "true":
        return None
    if not settings_enabled:
        return None
    if is_tty is None:
        is_tty = sys.stdout.isatty()
    if not is_tty:
        return None

    now = now or _now_utc()
    # When the caller does not pin a state path, route through the
    # fallback resolver so a pre-2.4 install's cadence file is still
    # consulted before we decide to re-check npm.
    read_path = state_path or _resolve_state_path()
    write_path = state_path or DEFAULT_STATE_PATH
    state = _read_state(read_path)
    if not _should_check(state, now):
        latest = state.get("last_seen_version")
        if isinstance(latest, str) and _is_newer(latest, installed_version):
            return _format_banner(latest, installed_version)
        return None

    latest = fetcher()
    payload = {
        "last_check_utc": now.strftime("%Y-%m-%dT%H:%M:%SZ"),
        "last_seen_version": latest or state.get("last_seen_version", ""),
        "installed_version": installed_version,
    }
    try:
        _write_state(write_path, payload)
    except OSError:
        pass

    if not latest or not _is_newer(latest, installed_version):
        return None
    return _format_banner(latest, installed_version)

File: src/scripts/_lib/update_check.ts
/**
 * Daily update-check banner for the `agent-config` dispatcher.
 *
 * TypeScript twin of `src/scripts/_lib/update_check.py` (ADR-200,
 * Phase 2 / Wave 1 batch B). Pure functions: `check_for_update()`
 * decides whether a banner should be emitted and returns the banner
 * string (or `null`). The dispatcher prints the returned string to
 * `stderr` after the subcommand finishes — never delaying the work,
 * never prompting.
 *
 * Design constraints (mirrored from the Python original):
 *
 * - Node builtins only (global `fetch` on Node >= 20).
 * - 1 s hard timeout on the registry call; network failure is silent.
 * - 24 h cadence gated by `~/.event4u/agent-config/update-check.json`
 *   (legacy `~/.config/agent-config/update-check.json` is read once as
 *   a fallback so the cadence is not reset on the first run after the
 *   namespace migration).
 * - Suppress in CI, on non-TTY stdout, when `AGENT_CONFIG_NO_UPDATE_CHECK=1`,
 *   or when `update_check.enabled: false` in settings.
 * - State file mode is `0600`.
 *
 * Port notes (intentional TS adaptations, semantics preserved):
 * - `fetch_latest_from_npm` / `check_for_update` are async — Node has
 *   no synchronous HTTP. The transport stays injectable via the
 *   `fetcher` parameter (same seam the pytest suite uses).
 */
import { randomBytes } from 'node:crypto';
import * as fs from 'node:fs';
import * as path from 'node:path';

import type { EnvMap } from './user_global_paths';
import * as user_global_paths from './user_global_paths';

export const PACKAGE_NAME = '@event4u/agent-config';
export const NPM_REGISTRY_URL = `https://registry.npmjs.org/${PACKAGE_NAME}/latest`;
/** Hard fetch timeout in seconds (mirrors the Python float). */
export const FETCH_TIMEOUT_S = 1.0;
/** 24 h cadence window in milliseconds (Python: `timedelta(hours=24)`). */
export const CHECK_WINDOW_MS = 24 * 60 * 60 * 1000;

export const STATE_FILENAME = 'update-check.json';

/**
 * Canonical write target. Reads are routed via `_resolve_state_path`
 * with a read-fallback to the legacy
 * `~/.config/agent-config/update-check.json` so a fresh install under
 * the new namespace does not lose the 24 h cadence window established
 * by a pre-2.4 install.
 *
 * NOTE: like the Python module-level constant, this is frozen at
 * import time (env override read once on module load) — replicated
 * quirk, not fixed.
 */
export const DEFAULT_STATE_PATH = user_global_paths.write_target(STATE_FILENAME);

/** Return the active state path, preferring the new namespace. */
function _resolve_state_path(): string {
    const found = user_global_paths.resolve_with_fallback(STATE_FILENAME);
    if (found !== null) {
        return found;
    }
    return DEFAULT_STATE_PATH;
}

export type Fetcher = () => string | null | Promise<string | null>;

/**
 * Return the `latest` dist-tag version, or `null` on any failure.
 *
 * Hard 1 s timeout. Any exception (network, JSON, missing key) yields
 * `null` — the update check is best-effort. `fetch_impl` is the
 * injectable transport (defaults to the global `fetch`).
 */
export async function fetch_latest_from_npm(
    options: {
        timeout?: number;
        url?: string;
        fetch_impl?: typeof fetch;
    } = {},
): Promise<string | null> {
    const timeout = options.timeout ?? FETCH_TIMEOUT_S;
    const url = options.url ?? NPM_REGISTRY_URL;
    const fetch_impl = options.fetch_impl ?? fetch;
    try {
        const resp = await fetch_impl(url, {
            headers: {
                Accept: 'application/json',
                'User-Agent': 'agent-config-update-check',
            },
            signal: AbortSignal.timeout(timeout * 1000),
        });
        const payload: unknown = await resp.json();
        if (payload !== null && typeof payload === 'object' && !Array.isArray(payload)) {
            const version = (payload as Record<string, unknown>).version;
            if (typeof version === 'string' && version.trim()) {
                return version.trim();
            }
        }
    } catch {
        return null;
    }
    return null;
}

function _read_state(state_path: string): Record<string, unknown> {
    try {
        const raw = fs.readFileSync(state_path, 'utf-8');
        const data: unknown = JSON.parse(raw);
        if (data !== null && typeof data === 'object' && !Array.isArray(data)) {
            return data as Record<string, unknown>;
        }
    } catch {
        // Unreadable / corrupt state — treat as empty, same as Python.
    }
    return {};
}

function _write_state(state_path: string, payload: Record<string, unknown>): void {
    const parent = path.dirname(state_path);
    fs.mkdirSync(parent, { recursive: true });
    // mkstemp equivalent: unique sibling temp file, created 0600.
    const tmp = path.join(parent, `.update-check-${randomBytes(8).toString('hex')}`);
    try {
        // json.dump(indent=2, sort_keys=True) equivalent. (Python also
        // escapes non-ASCII via ensure_ascii=True — flagged divergence;
        // payload values are version strings + timestamps, ASCII-only.)
        const sorted: Record<string, unknown> = {};
        for (const key of Object.keys(payload).sort()) {
            sorted[key] = payload[key];
        }
        fs.writeFileSync(tmp, JSON.stringify(sorted, null, 2), { encoding: 'utf-8', mode: 0o600 });
        fs.chmodSync(tmp, 0o600);
        fs.renameSync(tmp, state_path);
    } catch (err) {
        try {
            fs.unlinkSync(tmp);
        } catch {
            // Best-effort cleanup, mirroring the Python `except OSError: pass`.
        }
        throw err;
    }
}

function _should_check(state: Record<string, unknown>, now: Date): boolean {
    const last = state.last_check_utc;
    if (typeof last !== 'string') {
        return true;
    }
    // Python: datetime.fromisoformat(last.replace("Z", "+00:00")); a
    // naive timestamp (no offset) is assumed UTC. JS `Date.parse` would
    // treat offset-less strings as *local* time — normalize to UTC first.
    let normalized = last.replace(/Z/g, '+00:00');
    if (!/[+-]\d{2}:?\d{2}$/.test(normalized)) {
        normalized = `${normalized}+00:00`;
    }
    const last_ms = Date.parse(normalized);
    if (Number.isNaN(last_ms)) {
        return true;
    }
    return now.getTime() - last_ms >= CHECK_WINDOW_MS;
}

function _format_banner(latest: string, installed: string): string {
    return (
        `ℹ️  agent-config ${latest} available (you have ${installed}).\n` +
        `    Update: npx ${PACKAGE_NAME} update`
    );
}

/** Strict integer parse mirroring Python's `int(p)` (else 0). */
function _int_or_zero(part: string): number {
    const trimmed = part.trim();
    if (/^[+-]?\d+$/.test(trimmed)) {
        return parseInt(trimmed, 10);
    }
    return 0;
}

/**
 * Best-effort semver comparison — mirrors the Python `_is_newer`
 * (leading `v`s stripped, pre-release tail dropped, first three
 * numeric parts compared, non-numeric parts coerced to 0).
 *
 * Exported (despite the underscore) because the pytest suite accesses
 * `uc._is_newer`; the vitest port does the same.
 */
export function _is_newer(latest: string, installed: string): boolean {
    const parse = (v: string): [number, number, number] => {
        const core = v.replace(/^v+/, '').split('-')[0] ?? '';
        const parts = core.split('.');
        const out: number[] = [];
        for (const p of parts.slice(0, 3)) {
            out.push(_int_or_zero(p));
        }
        while (out.length < 3) {
            out.push(0);
        }
        return [out[0] as number, out[1] as number, out[2] as number];
    };

    const a = parse(latest);
    const b = parse(installed);
    for (let i = 0; i < 3; i += 1) {
        if ((a[i] as number) > (b[i] as number)) {
            return true;
        }
        if ((a[i] as number) < (b[i] as number)) {
            return false;
        }
    }
    return false;
}

/** Format a Date as `%Y-%m-%dT%H:%M:%SZ` (UTC, no milliseconds). */
function _format_utc(d: Date): string {
    return d.toISOString().replace(/\.\d{3}Z$/, 'Z');
}

/**
 * Decide whether to show an update banner. Pure (modulo state file).
 *
 * Returns the banner string or `null`. `null` covers every
 * suppression branch (CI, non-TTY, opt-out, within 24 h, network
 * failure, no update available).
 */
export async function check_for_update(
    installed_version: string,
    options: {
        now?: Date | null;
        state_path?: string | null;
        env?: EnvMap | null;
        is_tty?: boolean | null;
        settings_enabled?: boolean;
        fetcher?: Fetcher;
    } = {},
): Promise<string | null> {
    const env = options.env ?? process.env;
    if (env.AGENT_CONFIG_NO_UPDATE_CHECK === '1') {
        return null;
    }
    if (env.CI === '1' || env.CI === 'true' || env.GITHUB_ACTIONS === 'true') {
        return null;
    }
    const settings_enabled = options.settings_enabled ?? true;
    if (!settings_enabled) {
        return null;
    }
    let is_tty = options.is_tty ?? null;
    if (is_tty === null) {
        is_tty = Boolean(process.stdout.isTTY);
    }
    if (!is_tty) {
        return null;
    }

    const now = options.now ?? new Date();
    // When the caller does not pin a state path, route through the
    // fallback resolver so a pre-2.4 install's cadence file is still
    // consulted before we decide to re-check npm.
    const read_path = options.state_path ?? _resolve_state_path();
    const write_path = options.state_path ?? DEFAULT_STATE_PATH;
    const state = _read_state(read_path);
    if (!_should_check(state, now)) {
        const latest = state.last_seen_version;
        if (typeof latest === 'string' && _is_newer(latest, installed_version)) {
            return _format_banner(latest, installed_version);
        }
        return null;
    }

    const fetcher = options.fetcher ?? (() => fetch_latest_from_npm());
    const latest = await fetcher();
    const fallback_seen = Object.prototype.hasOwnProperty.call(state, 'last_seen_version')
        ? state.last_seen_version
        : '';
    const payload: Record<string, unknown> = {
        last_check_utc: _format_utc(now),
        // Python: `latest or state.get("last_seen_version", "")`.
        last_seen_version: latest ? latest : fallback_seen,
        installed_version,
    };
    try {
        _write_state(write_path, payload);
    } catch (err) {
        // Python swallows OSError only; everything else propagates.
        const code = (err as NodeJS.ErrnoException).code;
        if (typeof code !== 'string') {
            throw err;
        }
    }

    if (!latest || !_is_newer(latest, installed_version)) {
        return null;
    }
    return _format_banner(latest, installed_version);
}
